src/cgutils.cpp

```diff
@@ -3571,6 +3646,7 @@ static jl_cgval_t emit_new_struct(jl_codectx_t &ctx, jl_value_t *ty, size_t narg
                                 ->setOrdering(AtomicOrdering::Unordered);
                 }
                 else if (jl_is_uniontype(jtype)) {
+                    assert(!field_promotable);
```

Comment:
This doesn't seem to be ensured anywhere, but just arbitrarily asserted here?

Reply:
This branch seemed to be about fieldtypes, not return types?

Reply:
Here is an example of a failure this caused, for reference: https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/1c5858b_vs_1c5858b/SumTypes.primary.log

Reply:
I can add a explicit check here but it might just be a bandaid